### PR TITLE
Add search abilities to BinSearch

### PR DESCRIPTION
**Pull request by duramato**
- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

**Comment by duramato** [comment]
@labrys if you could test it i'd appreaciate, and if you @medariox could take a look at the code i'd appreciate too :P
Everyone one else is welcome to test/review too

**Comment by labrys** [comment]
@duramato running some other tests right now, but when those are done I'll test this

File: medusa/providers/nzb/binsearch.py [hunk 1]
```diff
@@ -121,7 +121,7 @@ def parse(self, data, mode):
                     title_re = title_regex.search(title)
                     # Try and get the the article subject from the wierd binsearch format
                     if title_re:
-                        title = title_re.group(1) or title_re.group(2) or attributes[2].find("span").get_text()
+                        title = title_re.group(1) if len(title_re.group(1)) > 3 or title_re.group(2) or attributes[2].find("span").get_text()
```
Comment (labrys) [inline medusa/providers/nzb/binsearch.py]: At this point this should probably be broken down into either separate steps or an if statement as its getting a little long

**Comment by duramato** [comment]
I'm facing some issues with the PR, i'm not understanding why the size isnt't working, and why the download link isn't either.

File: medusa/providers/nzb/binsearch.py [hunk 2]
```diff
@@ -54,6 +62,102 @@ def __init__(self):
...
+        with BS4Parser(data, 'html5lib') as html:
+            torrent_table = html.find('table', class_='xMenuT')
+            torrent_rows = torrent_table.find("tbody").find("tr") if torrent_table else []
+
+            if len(torrent_rows) < 1:
+                logger.log('Data returned from provider does not contain any torrents', logger.DEBUG)
+                return items
+
+            # Skip column headers
+            for row in torrent_table:
```
Comment (labrys) [inline medusa/providers/nzb/binsearch.py]: Shouldn't this be `for row in torrent_rows:`?

**Comment by labrys** [comment]
Needs final flake and testing, but should be ready to go now.

**Comment by duramato** [comment]
@labrys about the regex mine https://regex101.com/r/GwusYX/1 vs yours https://regex101.com/r/GwusYX/2

**Comment by duramato** [comment]
```
2017-04-24 10:51:15 INFO     SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Performing episode search for The Last Man on Earth
2017-04-24 10:51:15 DEBUG    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Search mode: Episode
2017-04-24 10:51:15 DEBUG    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Search string: The Last Man on Earth S03E15
2017-04-24 10:51:16 DEBUG    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] GET URL: https://www.binsearch.info/index.php?adv_age=&max=250&server=1&q=The+Last+Man+on+Earth+S03E15&xminsize=20 [Status: 200]
2017-04-24 10:51:16 DEBUG    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Traceback (most recent call last):
  File "/home/pi/Medusa/medusa/search/core.py", line 547, in search_providers
    down_cur_quality, manual_search, manual_search_type)
  File "/home/pi/Medusa/medusa/providers/generic_provider.py", line 211, in find_search_results
    items_list += self.search(search_string, ep_obj=episode)
  File "/home/pi/Medusa/medusa/providers/nzb/binsearch.py", line 91, in search
    results += self.parse(response.text, mode)
  File "/home/pi/Medusa/medusa/providers/nzb/binsearch.py", line 118, in parse
    headers = rows[0]('th')
IndexError: list index out of range

2017-04-24 10:51:16 ERROR    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Unknown exception while searching BinSearch. Error: 'list index out of range'
Traceback (most recent call last):
  File "/home/pi/Medusa/medusa/search/core.py", line 547, in search_providers
    down_cur_quality, manual_search, manual_search_type)
  File "/home/pi/Medusa/medusa/providers/generic_provider.py", line 211, in find_search_results
    items_list += self.search(search_string, ep_obj=episode)
  File "/home/pi/Medusa/medusa/providers/nzb/binsearch.py", line 91, in search
    results += self.parse(response.text, mode)
  File "/home/pi/Medusa/medusa/providers/nzb/binsearch.py", line 118, in parse
    headers = rows[0]('th')
IndexError: list index out of range
```

**Comment by duramato** [comment]
```
2017-04-24 11:05:05 INFO     SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Performing episode search for The Last Man on Earth
2017-04-24 11:05:05 DEBUG    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Search mode: Episode
2017-04-24 11:05:05 DEBUG    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Search string: The Last Man on Earth S03E01
2017-04-24 11:05:05 DEBUG    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] GET URL: https://www.binsearch.info/index.php?adv_age=&max=250&server=1&q=The+Last+Man+on+Earth+S03E01&xminsize=20 [Status: 200]
2017-04-24 11:05:06 DEBUG    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Found result: 160919_18 The.Last.Man.On.Earth.S03E01.General.Breast.Theme.with.Cobras.1080p.WE B-DL.DD5.1.H.264-BTN
2017-04-24 11:05:06 DEBUG    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Found result: 160919_17 The.Last.Man.On.Earth.S03E01.General.Breast.Theme.with.Cobras.720p.WEB -DL.DD5.1.H.264-BTN
2017-04-24 11:05:06 DEBUG    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Traceback (most recent call last):
  File "/home/pi/Medusa/medusa/search/core.py", line 547, in search_providers
    down_cur_quality, manual_search, manual_search_type)
  File "/home/pi/Medusa/medusa/providers/generic_provider.py", line 211, in find_search_results
    items_list += self.search(search_string, ep_obj=episode)
  File "/home/pi/Medusa/medusa/providers/nzb/binsearch.py", line 91, in search
    results += self.parse(response.text, mode)
  File "/home/pi/Medusa/medusa/providers/nzb/binsearch.py", line 131, in parse
    title = title_regex.search(title_field.text).group(1)
AttributeError: 'NoneType' object has no attribute 'group'

2017-04-24 11:05:06 ERROR    SEARCHQUEUE-MANUAL-281622 :: [BinSearch] :: [b99a467] Unknown exception while searching BinSearch. Error: "'NoneType' object has no attribute 'group'"
Traceback (most recent call last):
  File "/home/pi/Medusa/medusa/search/core.py", line 547, in search_providers
    down_cur_quality, manual_search, manual_search_type)
  File "/home/pi/Medusa/medusa/providers/generic_provider.py", line 211, in find_search_results
    items_list += self.search(search_string, ep_obj=episode)
  File "/home/pi/Medusa/medusa/providers/nzb/binsearch.py", line 91, in search
    results += self.parse(response.text, mode)
  File "/home/pi/Medusa/medusa/providers/nzb/binsearch.py", line 131, in parse
    title = title_regex.search(title_field.text).group(1)
AttributeError: 'NoneType' object has no attribute 'group'
```

File: medusa/providers/nzb/binsearch.py [hunk 3]
```diff
@@ -111,7 +111,7 @@ def process_column_header(td):
             table = html.find('table', class_='xMenuT')
             rows = table('tr') if table else []
             row_offset = 2
-            if not len(rows) - row_offset:
+            if not len(rows) - row_offset or not rows:
```
Comment (labrys) [inline medusa/providers/nzb/binsearch.py]: not rows should come first

File: medusa/providers/nzb/binsearch.py [hunk 4]
```diff
@@ -41,6 +52,230 @@ def __init__(self):
...
+                             {'result': result.name, 'location': filename})
+                    return True
+
+        if urls:
+            log.warning('Failed to download any results for {result}',
+                        {'result': result.name})
+
+        return False
+
+    def download_file(self, url, data, filename, session=None, headers=None, **kwargs):
```
Comment (p0psicles) [inline medusa/providers/nzb/binsearch.py]: Then only real change here, is adding the data paramater, and changing the request to a POST. The entire method can be removed if we add some params to the original download_files.

File: medusa/providers/nzb/binsearch.py [hunk 5]
```diff
@@ -41,6 +52,230 @@ def __init__(self):
...
+                    'pubdate': pubdate,
+                }
+                if mode != 'RSS':
+                    log.debug('Found result: {0}', title)
+
+                items.append(item)
+
+        return items
+
+    def download_result(self, result):
```
Comment (p0psicles) [inline medusa/providers/nzb/binsearch.py]: This is mostly the hacky part. we need to pass the nzb_id some way.

File: medusa/helpers/__init__.py [hunk 6]
```diff
@@ -1797,6 +1797,7 @@ def get_broken_providers():
         return []
 
     log.info('Broken providers found: {0}', response)
+    response = [u'sdbits', u'extratorrent', u'freshontv', u'bithdtv']
```
Comment (p0psicles) [inline medusa/helpers/__init__.py]: oow oeps. Don't forget to remove this line, before merging this.

File: medusa/providers/nzb/binsearch.py [hunk 7]
```diff
@@ -41,6 +52,230 @@ def __init__(self):
...
+                    title_field = col['subject'].find('span')
+                    # Try and get the the article subject from the weird binsearch format
+                    title = BinSearchProvider.title_regex.search(title_field.text).group(1)
+                    for extension in ('.nfo', '.par2', '.rar', '.zip', '.nzb'):
+                        # Strip extensions that aren't part of the file name
+                        title = title.rstrip(extension)
+                except AttributeError as error:
+                    log.debug('Parsing rows, that may not always have usefull info. Skipping to next.')
+                    continue
+                except Exception as error:
```
Comment (p0psicles) [inline medusa/providers/nzb/binsearch.py]: Ooh yeah, need to remove this also. Was to check if any other exceptions pass through.

**Comment by duramato** [comment]
@p0psicles 

image

Guess you only tested with blackhole?

**Comment by p0psicles** [comment]
Our current nzb client interface only supports sending the nzb url, not the file.
So as I see it the only solution for this would be:
* Add a new method to the nzb clients. For ex. send_nzb_file().
* Implement the POST/PUT in that method.
* Create an over writable method in NZBProvider, that makes the content/file ready for sending.

**Comment by fernandog** [comment]
@p0psicles @duramato nzbget supports sending the content:

```python
        # v13+ has a new combined append method that accepts both (url and
        # content) also the return value has changed from boolean to integer
        # (Positive number representing NZBID of the queue item. 0 and negative
        # numbers represent error codes.)
        elif nzbget_version >= 13:
            nzbget_result = nzbGetRPC.append(
                nzb.name + '.nzb',
                nzbcontent64 if nzbcontent64 is not None else nzb.url,
                category, nzbgetprio, False, False, dupekey, dupescore,
                'score'
            ) > 0
```

**Comment by fernandog** [comment]
SAB api allows it too but it not implemented. 

_Upload NZB using POST multipart/form-data. The file data should be in the field name.
For other parameters and output see addurl._ 
https://sabnzbd.org/wiki/advanced/api#addfile

**Comment by p0psicles** [comment]
@duramato should work now with sabnzbd (blackhole and api),
Would you want to do some testing?

It's rather hacked in currently. So next step will cleaning it up.
But at least binsearch should be usable now.

File: medusa/providers/nzb/binsearch.py [hunk 8]
```diff
@@ -6,120 +6,236 @@
...
+                print(size)
+                size = int(size)
+
+                query = query or title
+
+                download_url = 'https://www.binsearch.info/fcgi/nzb.fcgi?q={query}&max=100&adv_age=1100&server='.format(
+                    query=query
+                )
+
+                download_url = '{download_url}|nzb_id={nzb_id}'.format(download_url=download_url, nzb_id=nzb_id)
```
Comment (p0psicles) [inline medusa/providers/nzb/binsearch.py]: @medariox i'm now storing the url and the id (which is needed for the post) using a pipe. We could also add a field to the cache db's for this, but as its only one provider that needs this, I feel this is good enough for now.
What do you think?

Comment (medariox) [reply medusa/providers/nzb/binsearch.py]: Agreed, I think it's alright if it remains an exception (1-2 providers).

**Comment by p0psicles** [comment]
I still don't know why where doing this for a provider who is offline 90% of the time.

**Comment by p0psicles** [comment]
Just need to implement for nzbget. But then it's pretty much finished.

**Comment by fernandog** [comment]
@p0psicles nzbget we already can send content as I said here:

https://github.com/pymedusa/Medusa/pull/2657#issuecomment-315804697

File: medusa/helpers/__init__.py (hunk 6 above)
Comment (medariox) [reply medusa/helpers/__init__.py]: _Friendly reminder_ 😄

Comment (p0psicles) [reply medusa/helpers/__init__.py]: ok done

**Comment by p0psicles** [comment]
Approve pls?

**Comment by medariox** [comment]
Approved

**Comment by p0psicles** [comment]
Dramatic, where so close! After all those months!

**Pull request merged**